extensions/amp-a4a/0.1/cryptographic-validator.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AdResponseType,
+  Validator,
+  ValidatorResult,
+} from './amp-ad-type-defs';
+import {SignatureVerifier, VerificationStatus} from './signature-verifier';
+import {getAmpAdMetadata} from './amp-ad-utils';
+import {getDefaultBootstrapBaseUrl} from '../../../src/3p-frame';
+import {signingServerURLs} from '../../../ads/_a4a-config';
+import {user} from '../../../src/log';
+import {utf8Decode} from '../../../src/utils/bytes';
+
+export const SIGNATURE_VERIFIER_PROPERTY_NAME =
+    'AMP_FAST_FETCH_SIGNATURE_VERIFIER_';
+
+const TAG = 'amp-ad-render';
+
+export class CryptographicValidator extends Validator {
+  /** @param {!Window} win */
+  getSignatureVerifier_(win) {
+    return win[SIGNATURE_VERIFIER_PROPERTY_NAME] ||
+        (win[SIGNATURE_VERIFIER_PROPERTY_NAME] =
+         new SignatureVerifier(win, signingServerURLs));
+  }
+
+  /**
+   * @param {boolean} verificationSucceeded
+   * @param {!ArrayBuffer} bytes
+   * @return {!./amp-ad-type-defs.ValidatorOutput}
+   */
+  createOutput_(verificationSucceeded, bytes) {
+    const creativeData = {
+      creativeMetadata: this.getAmpAdMetadata_(utf8Decode(bytes)),
```

Comment:
Saw this https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/0.1/amp-a4a.js#L783. Not sure I fully understand, but do we want to add the check here as well?

Reply:
I think this concern is largely obsolete, and, in any case, shouldn't apply here.